dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/managedobject/CDIEJBManagedObjectFactoryImpl.java

```diff
@@ -165,19 +186,32 @@ private synchronized void doInitialization() {
                 }
             }
 
-            if (ejbDescriptor == null && partialMatchDescriptors.isEmpty()) {
-                Tr.error(tc, "Could not find an EjbDescriptor for : " + this.ejbName);
-            } else if (ejbDescriptor == null && ! partialMatchDescriptors.isEmpty()) {
-                String listStr = "[";
-                for (EjbDescriptor<T> desc : partialMatchDescriptors) {
-                    listStr += desc.getEjbName() + ",";
+            if (ejbDescriptor == null) {
+                if (firstDescriptor != null) {
+                    //We didn't find an ejbDescriptor thta pointed to the bean, but because there might have been a @Vetoed bean we record the first ejbDescriptor we found, assuming we found any.
+                    this.ejbDescriptor = firstDescriptor;
+                    this.ejbBeanManager = firstBeanManager;
+                    this.ejbBDA = firstBDA;
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "Setting ejbDescriptor to firstEjbDescriptor: " + firstDescriptor.toString());
+                    }
+                 } else {
+                    Tr.error(tc, "Could not find an EjbDescriptor for : " + this.ejbName);
+                 }
+            }
+
+            if (bean == null) {
+                //The bean can be null if it's Vetoed. 
+                Tr.error(tc, "Found the following EjbDescriptor " + this.ejbDescriptor + " for ejbName " + this.ejbName + " but it did not produce a bean or produced the wrong bean");
+                if (! partialMatchDescriptors.isEmpty()) {
+                    String listStr = "[";
+                    for (EjbDescriptor<T> desc : partialMatchDescriptors) {
+                        listStr += desc.getEjbName() + ",";
+                    }
+                    listStr = listStr.substring(0, listStr.length() - 1);
+                    listStr += "]";
+                    Tr.error(tc, "Found the following EjbDescriptors but they produced the wrong bean " + listStr);
```

Comment:
* If there are cases where this is expected, these log messages shouldn't be errors. Maybe a warning at most if it's usually wrong, but more likely an info message.
* Error/warning/info messages need to be translated
* It's a bit odd to have both `Tr.error` messages here. From the code I can see how they're saying different things but just reading the messages it's less clear. Maybe there's nothing we can do about this.